libvmi/arch/intel.c

```diff
@@ -88,33 +90,41 @@ uint64_t pdba_base_pae (uint64_t pdpe)
 }
 
 static inline
-uint32_t get_pgd_nopae (vmi_instance_t instance,
+status_t get_pgd_nopae (vmi_instance_t instance,
     uint32_t vaddr,
     uint32_t pdpe,
-    addr_t *pgd_entry)
+    addr_t *pgd_entry,
+    addr_t *pgd_value)
 {
-    uint32_t value;
+    uint32_t value = 0;
+    *pgd_value = 0;
     *pgd_entry = pdba_base_nopae(pdpe) + pgd_index_nopae(vaddr);
-    dbprint(VMI_DEBUG_PTLOOKUP, "--PTLookup: pgd_entry = 0x%.8"PRIx64"\n", *pgd_entry);
     if(VMI_FAILURE == vmi_read_32_pa(instance, *pgd_entry, &value)) {
-        value = 0;
+        dbprint(VMI_DEBUG_PTLOOKUP, "--PTLookup: failed to read pgd_entry at = 0x%.8"PRIx64"\n", *pgd_entry);
+        return VMI_FAILURE;
     }
-    return value;
+    *pgd_value = value;
```

Comment:
Since we only have one failure result, I think letting these fall through the the ENTRY_PRESENT checks is probably still fine.  If we were to do anything in the case that 0 is read, I'd say it should be an `errprint()` because it probably means the entry address was miscalculated or there is something unsupported by libvmi going on.